### PR TITLE
Added 'More' menu item when all items do not fit current resolution

### PR DESCRIPTION
**Pull request by jimleroyer**
# Summary | Résumé

[Trello card](https://trello.com/c/773NHG0t/333-nav-bug-more-plus-menu-option-when-there-are-many-items)

These changes add the *More* menu item on medium and larger viewports.

There are a few fixes accompanying this PR though:

* Partial move of the logic into the Notify Javascript extension mechanism. The Notify front-end has support for a `data-module` HTML attribute that specifies which piece of Javascript module should run on the decorated HTML element. This makes the code easier to test, reuse and maintain. I put all navigation logic into one new `NavMenu` module, but that isn't right because there are lots of unrelated logic (although all for navigation) in this new module. I made it this way for now to get to the next stage while improving just a bit. 
* Some fixes involving `undefined` variables when the JS uglifier did not run. This made the debugging of Javascript files in the local environment harder as we could not get it to run properly without the uglification process. Well that works now, we can debug locally and read the original Javascript files now.  

# Test instructions | Instructions pour tester la modification

**tl;dr:** Visit the homepage screens.

At the moment, only the non-logged-in homepage screens **in French** triggers the *More* menu item to appear (that I know of). Smaller resolutions will trigger the smaller viewports before the *More* menu item appears; these don't have the horizontal menu items so the *More* menu item is unnecessary. 

# Design // SS

The following SS comes from [the figma design](https://www.figma.com/file/r3GzRBjBDWvL2AwCWq1e1W/Navigating-the-site-%7C-Naviguer-le-site?node-id=5724%3A967):

image

The following screenshot comes from the actual implementation:

image

I was not sure of the *More* menu item behaviour so I made it an overlay menu for now. These are the differences:

* It should push down the content, it should create a white space between the menu and content for every menu items displayed vertically. 
* The alignment with the `More` menu item and its sub-elements is a bit off. This is hard to align in an exact similar manner. I can see a way to do it but with more work. 

# Reviewer checklist | Liste de vérification du réviseur

This is a suggested checklist of questions reviewers might ask during their review | Voici une suggestion de liste de vérification comprenant des questions que les réviseurs pourraient poser pendant leur examen :

- [ ] Does this meet a user need? | Est-ce que ça répond à un besoin utilisateur?
- [ ] Is it accessible? | Est-ce que c’est accessible?
- [ ] Is it translated between both offical languages? | Est-ce dans les deux langues officielles?
- [ ] Is the code maintainable? | Est-ce que le code peut être maintenu?
- [ ] Have you tested it? | L’avez-vous testé?
- [ ] Are there automated tests? | Y a-t-il des tests automatisés?
- [ ] Does this cause automated test coverage to drop? | Est-ce que ça entraîne une baisse de la quantité de code couvert par les tests automatisés?
- [ ] Does this break existing functionality? | Est-ce que ça brise une fonctionnalité existante?
- [ ] Should this be split into smaller PRs to decrease change risk? | Est-ce que ça devrait être divisé en de plus petites demandes de tirage (« pull requests ») afin de réduire le risque lié aux modifications?
- [ ] Does this change the privacy policy? | Est-ce que ça entraîne une modification de la politique de confidentialité?
- [ ] Does this introduce any security concerns? | Est-ce que ça introduit des préoccupations liées à la sécurité?
- [ ] Does this significantly alter performance? | Est-ce que ça modifie de façon importante la performance?
- [ ] What is the risk level of using added dependencies? | Quel est le degré de risque d’utiliser des dépendances ajoutées?
- [ ] Should any documentation be updated as a result of this? (i.e. README setup, etc.) | Faudra-t-il mettre à jour la documentation à la suite de ce changement (fichier README, etc.)?

### CODE REVIEW

File: app/assets/javascripts/navMenu.js [hunk 1]
```diff
@@ -0,0 +1,183 @@
+(function(Modules) {
```
Comment (jimleroyer) [inline app/assets/javascripts/navMenu.js:1]: This Notify JS module can be used with `data-attribute="nav-menu"` on an HTML element to trigger the Javascript below.

File: app/assets/javascripts/navMenu.js [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+      moreMenuItems.children().removeClass("text-center");
+      moreMenuItems.children().addClass("text-right");
+    } else {
+      // Hide the More menu when it does not contain item(s).
+      moreMenu.addClass("hidden");
+    }
+  }
+
+  Modules.NavMenu = function() {
+    this.start = function(component) {
```
Comment (jimleroyer) [inline app/assets/javascripts/navMenu.js:178]: The `component` parameter is the HTML element decorated with the `data-attribute="nav-menu"` attribute.

Comment (jimleroyer) [reply app/assets/javascripts/navMenu.js:178]: At the moment, it is not ideal because this module deals with 4 different menus (i.e. desktop nav menu, mobile nav menu, account desktop menu & account mobile menu). There could be one module for each of these, isolated from each other, along with potential generic modules, for example: scoping tabbing navigation for accessibility purposes (a simple `data-attribute` on an HTML element would add the functionality on any other element).

File: app/templates/admin_template.html [hunk 3]
```diff
@@ -22,8 +22,3 @@
     </div>
   </div>
 {% endblock %}
-
-{% block body_end %}
-  <script src="{{ asset_path }}javascripts/main.min.js"></script>
-  <script type="text/javascript" src="{{ asset_url('javascripts/all.min.js') }}"></script>
-{% endblock %}
```
Comment (jimleroyer) [inline app/templates/admin_template.html:29]: The Javascript files were exclusive to the logged-in admin template before. I moved it to be available on other pages now.

File: app/templates/main_template.html [hunk 4]
```diff
@@ -152,6 +152,9 @@
     )
 </script>
 
-{% block page_script %}{% endblock %}
+{% block page_script %}
+<script src="{{ asset_path }}javascripts/main.min.js"></script>
+<script type="text/javascript" src="{{ asset_url('javascripts/all.min.js') }}"></script>
+{% endblock %}
```
Comment (jimleroyer) [inline app/templates/main_template.html:158]: Notice the override of the `page_script` template block. We did not use this intended block before.

File: app/templates/notify_template.html [hunk 5]
```diff
@@ -22,5 +22,7 @@
   </head>
   <body>
     {% block main_content %}{% endblock %}
+    <script src="{{ asset_path }}javascripts/main.min.js"></script>
+    <script type="text/javascript" src="{{ asset_url('javascripts/all.min.js') }}"></script>
```
Comment (jimleroyer) [inline app/templates/notify_template.html:26]: This adds the Javascript files to the homepage as well.

File: app/templates/partials/nav/gc_header_nav.html [hunk 6]
```diff
@@ -12,7 +12,7 @@
     <!-- Mobile menu and nav bar -->
     <div class="grid grid-cols-2">
         <div class="py-2">
-          <a href="#" id="menu"
+          <a href="#" id="menu" data-module="nav-menu"
```
Comment (jimleroyer) [inline app/templates/partials/nav/gc_header_nav.html:15]: The Notify js extension mechanism is triggered with this attribute.

File: app/templates/partials/nav/gc_header_nav.html [hunk 7]
```diff
@@ -102,106 +102,27 @@
...
-
-  safeAddEventListener("#account-icon-mobile", "click", accountMobileToggle)
-  safeAddEventListener("#account-close-mobile", "click", accountMobileToggle)
-  registerMenuEscape("#account-menu-mobile", accountMobileToggle)
-
-  safeAddEventListener("#account-menu", "click", accountMenuToggle)
-  registerMenuEscape("#account-menu-options", accountMenuToggle)
-  registerTabNavigationScope('#account-menu-mobile')
-
-</script>
```
Comment (jimleroyer) [inline app/templates/partials/nav/gc_header_nav.html:207]: All this Javascript will now get into the regular JS build pipeline and aligned with Notify Javascript way of doing.

File: app/translations/csv/fr.csv [hunk 8]
```diff
@@ -1300,3 +1300,4 @@
 "Daily sending limit","Limite d’envoi quotidienne"
 "{} notifications","{} notifications"
 "{} calls","{} appels"
+"More","Plus"
```
Comment (jimleroyer) [inline app/translations/csv/fr.csv:1303]: Do we really need this? I heard before that *Less is more*. 🤔 💭

File: app/assets/javascripts/index.js [hunk 9]
```diff
@@ -8,11 +8,22 @@ import { App } from "./scheduler/App";
...
 
+if (!window.APP_PHRASES || typeof APP_PHRASES === "undefined") {
+  window.APP_PHRASES = {
+    now: "Now"
+  };
+}
+
+if (!window.APP_LANG || typeof APP_LANG === "undefined") {
+  window.APP_LANG = "en";
+}
```
Comment (AntoineAugusti) [inline app/assets/javascripts/index.js:19]: Why is this needed? Is this code not enough https://github.com/cds-snc/notification-admin/blob/dfe53538dd49f40f009162793578de2bd347450e/app/templates/main_template.html#L40-L41?

File: app/assets/javascripts/navMenu.js (hunk 1 above)
Comment (AntoineAugusti) [reply app/assets/javascripts/navMenu.js:1]: Can you point out what's new in this file? I understand you moved it here from the HTML template. This would help understand what to pay attention to.

File: app/assets/javascripts/navMenu.js [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+    safeAddEventListener("#menu", "click", menuMobileToggle)
+    registerMenuEscape(".mobile-nav", menuMobileToggle)
+
+    safeAddEventListener("#account-icon-mobile", "click", accountMobileToggle)
+    safeAddEventListener("#account-close-mobile", "click", accountMobileToggle)
+    registerMenuEscape("#account-menu-mobile", accountMobileToggle)
+
+    safeAddEventListener("#account-menu", "click", accountMenuToggle)
+    registerMenuEscape("#account-menu-options", accountMenuToggle)
+    registerTabNavigationScope('#account-menu-mobile')
```
Comment (AntoineAugusti) [inline app/assets/javascripts/navMenu.js:107]: I've seen teams prefixing HTML IDs used only for JS purposes with `js-` to make sure it's clear that it's used for this purpose and it could break if deleted. WDYT?

File: app/templates/main_template.html [hunk 11]
```diff
@@ -152,6 +152,9 @@
     )
 </script>
 
-{% block page_script %}{% endblock %}
+{% block page_script %}
+<script src="{{ asset_path }}javascripts/main.min.js"></script>
```
Comment (AntoineAugusti) [inline app/templates/main_template.html:156]: Could use `asset_url` here to leverage fingerprinting?

File: app/translations/csv/fr.csv (hunk 8 above)
Comment (AntoineAugusti) [reply app/translations/csv/fr.csv:1303]: I think it's helpful to highlight that there is indeed _more_. Like adding "Menu" on top of the "hamburger icon" to be explicit.

File: app/templates/notify_template.html [hunk 12]
```diff
@@ -22,5 +22,7 @@
   </head>
   <body>
     {% block main_content %}{% endblock %}
+    <script src="{{ asset_path }}javascripts/main.min.js"></script>
```
Comment (AntoineAugusti) [inline app/templates/notify_template.html:25]: Could use `asset_url` here to leverage fingerprinting?

File: app/assets/javascripts/navMenu.js (hunk 10 above)
Comment (jimleroyer) [reply app/assets/javascripts/navMenu.js:107]: I agree, sounds like a good practice. With my upcoming PR, we should not have these explicitly referred IDs though.

File: app/templates/main_template.html (hunk 11 above)
Comment (jimleroyer) [reply app/templates/main_template.html:156]: Sounds good, let me try if that does not give some trouble before.

File: app/assets/javascripts/index.js (hunk 9 above)
Comment (jimleroyer) [reply app/assets/javascripts/index.js:19]: It produces an error if ran locally and without the uglification process. I saw that same logic in other files as well and it is my assumption at the moment that there is no priority ordering of js module execution guaranteed. Hence, many similar code excerpts can be found to assume that previous code wasn't ran. I would validate this theory and consider this a potential tech debt.

File: app/assets/javascripts/navMenu.js (hunk 1 above)
Comment (jimleroyer) [reply app/assets/javascripts/navMenu.js:1]: Yes let me add more comments to say what is new.

File: app/assets/javascripts/navMenu.js [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+
+  function accountMenuToggle() {
+    const accountMenu = document.querySelector("#account-menu-options")
+    const arrow = document.querySelector(".account-menu-arrow")
+    accountMenu.classList.toggle("hidden")
+    arrow.classList.toggle("flip")
+    accountMenu.querySelector('a').focus()
+  }
+
+  function init($menu) {
```
Comment (jimleroyer) [inline app/assets/javascripts/navMenu.js:97]: This function is new.

File: app/assets/javascripts/navMenu.js [hunk 14]
```diff
@@ -0,0 +1,183 @@
...
+      ctx.varWidth += $(this).outerWidth();
+    });
+
+    resizeMenu(ctx); 
+    $(window).resize(function() {
+      resizeMenu(ctx);
+    });
+  }
+
+  function resizeMenu(ctx) {
```
Comment (jimleroyer) [inline app/assets/javascripts/navMenu.js:137]: This function is new.

File: app/assets/javascripts/navMenu.js [hunk 15]
```diff
@@ -0,0 +1,183 @@
...
+      moreMenu.css({"display": moreMenuDisplay, "width": moreMenuWidth});
+      moreMenuItems.children().removeClass("text-center");
+      moreMenuItems.children().addClass("text-right");
+    } else {
+      // Hide the More menu when it does not contain item(s).
+      moreMenu.addClass("hidden");
+    }
+  }
+
+  Modules.NavMenu = function() {
```
Comment (jimleroyer) [inline app/assets/javascripts/navMenu.js:177]: The module setup is new.

File: app/assets/javascripts/mobileMenuAccount.js [hunk 16]
```diff
@@ -0,0 +1,64 @@
+(function (Modules) {
+  "use strict";
+
+  function registerMenuEscape($selector, toggleFn) {
```
Comment (QuanNguyen-CDS) [inline app/assets/javascripts/mobileMenuAccount.js:4]: I see this code also being in menu.js, can we combine them (following the DRY principles)?

File: app/assets/javascripts/moreMenu.js [hunk 17]
```diff
@@ -0,0 +1,94 @@
...
+    });
+  }
+
+  function resizeMenu() {
+    // Current available width of the menu.
+    elemWidth = $menuItems.width();
+
+    // Calculate how many items fit on the total width.
+    // Substract 1 as CSS :gt property is 0-based index.
+    fitCount = Math.floor((elemWidth / varWidth) * ctr) - 1;
```
Comment (QuanNguyen-CDS) [inline app/assets/javascripts/moreMenu.js:47]: could varWidth be zero here (ever?) - could we run into a div-by-zero scenario?

File: app/assets/javascripts/scheduler/store.js [hunk 18]
```diff
@@ -59,6 +59,7 @@ export const defaultState = (
     updateMessage: "",
     _24hr: LOCALE === "en" ? "off" : "on",
     errors: "",
+    // REVIEW: Raises error past 23h00m in a day, not great when debugging late.
```
Comment (QuanNguyen-CDS) [inline app/assets/javascripts/scheduler/store.js:62]: 👍

File: app/assets/javascripts/scopeTabNavigation.js [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+   * other elements of the page as long as the container is not reached. This
+   * works well within a modal popup that took focus away from the main window.
+   */
+  function registerTabNavigationScope(modal) {
+    const firstFocusableElement = modal.querySelectorAll(focusableElements)[0];
+    const focusableContent = modal.querySelectorAll(focusableElements);
+    const lastFocusableElement = focusableContent[focusableContent.length - 1];
+
+    document.addEventListener("keydown", function (e) {
+      let isTabPressed = e.key === "Tab" || e.keyCode === 9;
```
Comment (QuanNguyen-CDS) [inline app/assets/javascripts/scopeTabNavigation.js:27]: From my understanding the use of keyCode is no longer recommended (https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/keyCode)? From the Mozilla reference, the recommendation is to use `KeyboardEvent.code` however it also states that some browser still do not have support for this either...

Not sure if there's a better alternative.

File: app/templates/partials/nav/gc_header_nav.html [hunk 20]
```diff
@@ -102,106 +106,30 @@
...
+          data-module-menu-items="proposition-links" 
+          data-module-menu-container="more-menu-container"
+          class="flex flex-shrink-0 text-center p-5 bg-gray hidden">
+          <a
+            href="#"
+            aria-hidden="true"
+            class="text-smaller text-blue visited:text-blue link:text-blue font-bold no-underline focus:outline-none">
+            {{ _('More') }}
+            <img
+              alt="Account menu expansion arrow"
```
Comment (QuanNguyen-CDS) [inline app/templates/partials/nav/gc_header_nav.html:121]: question: is this page language aware? do we need to consider `alt` in both languages [French, English]?

**Comment by anikbrazeau** [comment]
This is starting to take shape! Encountered it a few times in signed in (with platform admin) as well 😄 

Wondering how much work it'd be to _not_ push down the content instead, when the `More` menu opens? Worried it might not be the behaviour users expect on desktop.

File: app/assets/javascripts/menu.js [hunk 21]
```diff
@@ -0,0 +1,65 @@
+(function (Modules) {
+  "use strict";
+
+  function registerMenuEscape($selector, toggleFn) {
+    $selector.keydown(function (e) {
```
[the remark this answers is not in the document]
Comment (jimleroyer) [reply app/assets/javascripts/menu.js:5]: It should trigger on key down and will only happen once.

File: app/assets/javascripts/mobileMenuAccount.js (hunk 16 above)
Comment (jimleroyer) [reply app/assets/javascripts/mobileMenuAccount.js:4]: Let me try to put this in a utils file.

My initial hope was to have the escape key registration as a reusable module that can stand on its own... but current module support can't let me define several of these on the same HTML element.

File: app/assets/javascripts/moreMenu.js (hunk 17 above)
Comment (jimleroyer) [reply app/assets/javascripts/moreMenu.js:47]: If there are no children, i.e. menu items. This should not happen.. but I can add an extra safe precaution.

File: app/assets/javascripts/scopeTabNavigation.js (hunk 19 above)
Comment (jimleroyer) [reply app/assets/javascripts/scopeTabNavigation.js:27]: Yeah, it's not supported by IE11, which we have to support AFAIK, although I'd be more than happy to drop this hopefully soon. Using polyfills could be a workaround such [as this one](https://www.npmjs.com/package/keyboardevent-key-polyfill). We could have an API front-end of our own for these.

File: app/templates/partials/nav/gc_header_nav.html (hunk 20 above)
Comment (jimleroyer) [reply app/templates/partials/nav/gc_header_nav.html:121]: Yes we need to. 😓 

Let me handle this in a separate PR though. I would like designer's review on our hardcoded alt images to know if these are proper description as they are.

**Comment by jimleroyer** [comment]
There is now a) a divider between items, b) a vertical sideline on selected page, c) removal of the horizontal line on selected page, d) the menu overlay on top of the admin section now, e) coding related PR comments have been addressed too.

There have been no adjustment in spacing/alignment.

**Comment by jimleroyer** [comment]
[NAV Bug: "More" / "Plus" menu option when there are many items](https://trello.com/c/773NHG0t/333-nav-bug-more-plus-menu-option-when-there-are-many-items)

File: app/templates/notify_template.html [hunk 22]
```diff
@@ -22,5 +22,8 @@
...
```
Comment (AntoineAugusti) [inline app/templates/notify_template.html:25]:
```suggestion
    <script src="{{ asset_path }}javascripts/main.min.js"></script>
```
You already include this file afterwards, I think you can delete that

**Review by AntoineAugusti — APPROVED** [review]
LGTM, a quick comment to address before merging 🚢

File: app/templates/notify_template.html (hunk 22 above)
Comment (jimleroyer) [reply app/templates/notify_template.html:25]: My mistake, I'll remove.
